src/testnet/node.rs

```diff
@@ -0,0 +1,385 @@
+        let registered_key = match &self.active_registered_key {
+            None => {
+                // We're continuously registering new keys, as such, this branch
+                // should be unreachable.
+                unreachable!()
+            },
+            Some(ref key) => key,
+        };
+
+        // Generates a proof out of the sortition hash provided in the params.
+        let vrf_proof = self.keychain.generate_proof(
+            &registered_key.vrf_public_key, 
+            sortitioned_block.sortition_hash.as_bytes()).unwrap();
+
+        // Generates a new secret key for signing the trail of microblocks
+        // of the upcoming tenure.
+        let microblock_secret_key = self.keychain.rotate_microblock_keypair();
+
+        // Get the stack's chain tip
+        let chain_tip = match self.bootstraping_chain {
+            true => StacksHeaderInfo::genesis(),
+            false => match &self.chain_tip {
+                Some(chain_tip) => chain_tip.clone(),
+                None => unreachable!()
+            }
+        };
+
+        // Constructs the coinbase transaction - 1st txn that should be handled and included in 
+        // the upcoming tenure.
+        let coinbase_tx = self.generate_coinbase_tx();
+
+        // Construct the upcoming tenure
+        let tenure = LeaderTenure::new(
+            chain_tip, 
+            self.average_block_time,
+            coinbase_tx,
+            self.config.clone(),
+            self.mem_pool.clone(),
+            microblock_secret_key, 
+            sortitioned_block.clone(),
+            vrf_proof);
+
+        Some(tenure)
+    }
+
+    /// Handles artefacts coming from an ongoing tenure.
+    /// At this point, we're not updating the chainstate, we're simply having the node
+    /// candidating for the next tenure.
+    pub fn receive_tenure_artefacts(&mut self, anchored_block_from_ongoing_tenure: &StacksBlock, parent_block: &SortitionedBlock) {
+        let ops_tx = self.burnchain_ops_tx.take().unwrap();
+
+        if self.active_registered_key.is_some() {
+            let registered_key = self.active_registered_key.clone().unwrap();
+
+            let vrf_proof = self.keychain.generate_proof(
+                &registered_key.vrf_public_key, 
+                parent_block.sortition_hash.as_bytes()).unwrap();
+
+            let op = self.generate_block_commit_op(
+                anchored_block_from_ongoing_tenure.header.block_hash(),
+                1, // todo(ludo): fix
+                &registered_key, 
+                &parent_block,
+                VRFSeed::from_proof(&vrf_proof));
+
+            ops_tx.send(op).unwrap();
+        }
+
+        // Naive implementation: we keep registering new keys
+        let vrf_pk = self.keychain.rotate_vrf_keypair();
+        let op = self.generate_leader_key_register_op(vrf_pk, parent_block.consensus_hash); // todo(ludo): should we use the consensus hash from the burnchain tip, or last sortitioned block?
```

Comment:
Use the one from the burn chain tip.  There's a consensus hash calculated for each block, even if the block has no sortition.

Reply:
Addressed with https://github.com/blockstack/blockstack-core/pull/1202/commits/d27608ec158350f7e4f000c5462fee090e92a851